File: 0523-continuous-subarray-sum/0523-continuous-subarray-sum.py
class Solution(object):
    def checkSubarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        prefixSum = [0]*len(nums)
        prefixSum[0] = nums[0];
        mp = {prefixSum[0]%k: 0}; # mod -> index
        for i in range(1,len(nums)):
            prefixSum[i] = prefixSum[i-1]+nums[i];
        for i in range(1,len(prefixSum)):
            if(prefixSum[i]%k==0):
                return True
            mod = prefixSum[i]%k
            if(mod in mp and i - mp[mod] > 1):
                return True
            if(mod not in mp):
                mp[mod] = i
        return False

File: 0523-continuous-subarray-sum/0523-continuous-subarray-sum.cpp
// [5,0,0,0]
// [5,5,5,5]
// mp = 2,0 
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> mp;
        vector<int> prefixSum(nums.size(), 0);
        prefixSum[0] = nums[0];
        for(int i = 1; i<nums.size(); i++){
            prefixSum[i] = prefixSum[i-1] + nums[i];
        }
        mp.insert({prefixSum[0]%k,0});
        for(int i = 1; i<prefixSum.size(); i++){
            if(prefixSum[i]%k==0) return true;
            int mod = prefixSum[i]%k;
            if(mp.count(mod) && i - mp[mod] > 1) return true;
            if(!mp.count(mod)){
                mp[mod] = i;
            }
        }
        return false;
    }
};